File: src/examples/tinybc/tinybc.cpp
#include "kali/kali.h"
#include "kali/core/platform.h"
#include "kali/core/bitmap.h"
#include "kali/core/timer.h"
#include "kali/device/device.h"
#include "kali/device/shader.h"
#include "kali/device/command.h"
#include "kali/device/shader_cursor.h"
#include "kali/device/shader_object.h"
#include "kali/device/kernel.h"
#include "kali/device/query.h"
#include "kali/device/agility_sdk.h"
#include "kali/utils/tev.h"

#include <argparse/argparse.hpp>

KALI_EXPORT_AGILITY_SDK

static const std::filesystem::path EXAMPLE_DIR(KALI_EXAMPLE_DIR);

using namespace kali;

int main(int argc, const char* argv[])
{
    argparse::ArgumentParser args("tinybc");
    args.add_description("Slang-based BC7 - mode 6 compressor");
    args.add_argument("input_path").required().help("Path to the input texture.");
    args.add_argument("-o", "--output_path").help("Optional path to save the decoded BC7 texture.");
    args.add_argument("-s", "--opt_steps").default_value(100).help("Number of optimization (gradient descene) steps.");
    args.add_argument("-b", "--benchmark")
        .default_value(false)
        .implicit_value(true)
        .help("Run in benchmark mode to measure processing time.");
    args.add_argument("-t", "--tev").default_value(false).implicit_value(true).help("Show images in tev image viewer.");
    args.add_argument("-v", "--verbose").default_value(false).implicit_value(true).help("Enable verbose logging.");

    try {
        // args.parse_args(argc, argv);
        KALI_UNUSED(argc, argv);
        args.parse_args(
            {"tinybc", "C:/projects/kali/monalisa.jpg", "-o", "C:/projects/kali/monalisa_bc7.jpg", "-b", "-v"}
        );
    } catch (const std::runtime_error& err) {
        std::cerr << err.what() << std::endl;
        return 1;
    }

    std::string input_path = args.get("input_path");
    std::optional<std::string> output_path = args.present("output_path");
    int opt_steps = args.get<int>("opt_steps");
    bool benchmark = args.get<bool>("benchmark");
    bool tev = args.get<bool>("tev");
    bool verbose = args.get<bool>("verbose");

    kali::static_init();

    ref<Bitmap> input;

    try {
        input = make_ref<Bitmap>(input_path)->convert(Bitmap::PixelFormat::rgba, Bitmap::ComponentType::float32, false);
    } catch (const std::exception& e) {
        fmt::println("Failed to load input image from {}:\n{}", input_path, e.what());
        return 1;
    }

    uint32_t w = input->width();
    uint32_t h = input->height();

    ref<Device> device = Device::create({.enable_debug_layers = verbose});

    // Create input texture
    ref<Texture> input_tex = device->create_texture({
        .type = TextureType::texture_2d,
        .format = Format::rgba32_float,
        .width = w,
        .height = h,
        .mip_count = 1,
        .usage = ResourceUsage::shader_resource,
    });
    // TODO use convenience function (or init data) when available
    {
        ref<CommandBuffer> command_buffer = device->create_command_buffer();
        command_buffer->upload_texture_data(input_tex, 0, input->data());
        command_buffer->submit();
    }

    // Show input texture in tev
    if (tev)
        utils::show_in_tev_async(input_tex, "tinybc-input");

    // Create decoded texture
    ref<Texture> decoded_tex = device->create_texture({
        .type = TextureType::texture_2d,
        .format = Format::rgba32_float,
        .width = w,
        .height = h,
        .mip_count = 1,
        .usage = ResourceUsage::unordered_access,
    });

    ref<ComputeKernel> encoder = device
                                     ->load_module(
                                         EXAMPLE_DIR / "tinybc.slang",
                                         DefineList({
                                             {"CONFIG_USE_ADAM", "true"},
                                             {"CONFIG_OPT_STEPS", std::to_string(opt_steps)},
                                         })
                                     )
                                     ->create_compute_kernel("main");

    uint32_t num_iters = benchmark ? 1000 : 1;

    Timer t;

    // Setup query pool to measure GPU time
    ref<QueryPool> queries = device->create_query_pool({.type = QueryType::timestamp, .count = num_iters * 2});

    // Compress!
    ref<CommandBuffer> command_buffer = device->create_command_buffer();
    for (uint32_t i = 0; i < num_iters; ++i) {
        command_buffer->write_timestamp(queries, i * 2);
        encoder->dispatch(
            uint3(w, h, 1),
            [&](ShaderCursor cursor)
            {
                cursor["input_tex"] = input_tex;
                cursor["decoded_tex"] = decoded_tex;
                cursor["lr"] = 0.1f;
                cursor["adam_beta_1"] = 0.9f;
                cursor["adam_beta_2"] = 0.999f;
            },
            command_buffer
        );
        command_buffer->write_timestamp(queries, i * 2 + 1);
    }
    command_buffer->submit();

    device->wait();

    double total_cpu_time_sec = t.elapsed_s();

    std::vector<double> times = queries->get_timestamp_results(0, num_iters * 2);
    double comp_time_sec = 0.0;
    for (uint32_t i = 0; i < num_iters; ++i)
        comp_time_sec += (times[i * 2 + 1] - times[i * 2]);
    comp_time_sec /= num_iters;

    // Calculate and print performance metrics
    if (benchmark) {
        double textures_per_sec = 1.0 / comp_time_sec;
        double giga_texels_per_sec = w * h * textures_per_sec / 1e9;
        fmt::println("Benchmark:");
        fmt::println("- Number of optimization steps: {}", opt_steps);
        fmt::println("- Compression time: {:.4g} ms", comp_time_sec * 1e3);
        fmt::println("- Compression throughput: {:.4g} GTexels/s", giga_texels_per_sec);
        fmt::println("- Total CPU time: {:.4g} s", total_cpu_time_sec);
    }

    // Calculate and print PSNR
    ref<Bitmap> decoded = decoded_tex->to_bitmap();
    double mse = 0.0;
    const float* input_data = input->data_as<float>();
    const float* decoded_data = decoded->data_as<float>();
    for (uint32_t i = 0; i < w * h * 4; ++i)
        mse += (input_data[i] - decoded_data[i]) * (input_data[i] - decoded_data[i]);
    mse /= w * h * 4;
    double psnr = 20.0 * std::log10(1.0 / std::sqrt(mse));
    fmt::println("PSNR: {:.4g}", psnr);

    // Show decoded texture in tev
    if (tev)
        utils::show_in_tev_async(decoded_tex, "tinybc-decoded");

    // Output decoded texture
    if (output_path) {
        decoded_tex->to_bitmap()
            ->convert(Bitmap::PixelFormat::rgb, Bitmap::ComponentType::uint8, true)
            ->write_async(*output_path);
    }

    kali::static_shutdown();

    return 0;
}

File: src/examples/tinybc/tinybc.py
# SPDX-License-Identifier: Apache-2.0

import sys
import argparse
import numpy as np
import kali
from pathlib import Path

EXAMPLE_DIR = Path(__file__).parent

# Build and parse command line
# fmt: off
parser = argparse.ArgumentParser(description="Slang-based BC7 - mode 6 compressor")
parser.add_argument("input_path", help="Path to the input texture.")
parser.add_argument("-o", "--output_path", help="Optional path to save the decoded BC7 texture.")
parser.add_argument("-s", "--opt_steps", type=int, default=100, help="Number of optimization (gradient descene) steps.")
parser.add_argument("-b", "--benchmark", action="store_true", help="Run in benchmark mode to measure processing time.")
parser.add_argument("-t", "--tev", action="store_true", help="Show images in tev image viewer.")
parser.add_argument("-v", "--verbose", action="store_true", help="Enable verbose logging.")
# fmt: on
# args = parser.parse_args()
args = parser.parse_args(
    [
        "C:/projects/kali/data/test_images/monalisa.jpg",
        "-o",
        "monalisa_bc7.jpg",
        "-t",
        "-b",
        "-v",
    ]
)

# Load input image
try:
    image = kali.Bitmap(args.input_path).convert(
        pixel_format=kali.Bitmap.PixelFormat.rgba,
        component_type=kali.Bitmap.ComponentType.float32,
        srgb_gamma=False,
    )
    w, h = image.width, image.height
    input = np.asarray(image).clip(0, 1)
except Exception as e:
    print(f"Failed to load input image from {args.input_path}:\n{e}")
    sys.exit(1)

# Create device
device = kali.Device(
    enable_debug_layers=args.verbose, compiler_options={"include_paths": [EXAMPLE_DIR]}
)

# Create input texture
input_tex = device.create_texture(
    format=kali.Format.rgba32_float,
    width=w,
    height=h,
    mip_count=1,
    usage=kali.ResourceUsage.shader_resource,
)
input_tex.from_numpy(input)

# Show input texture in tev
if args.tev:
    kali.utils.show_in_tev_async(input_tex, name="tinybc-input")

# Create decoded texture
decoded_tex = device.create_texture(
    format=kali.Format.rgba32_float,
    width=w,
    height=h,
    mip_count=1,
    usage=kali.ResourceUsage.unordered_access,
)

# Load shader module
constants = f"export static const bool USE_ADAM = true;\nexport static const uint OPT_STEPS = {args.opt_steps};\n"
program = device.load_program("tinybc.slang", ["main"], constants)
encoder = device.create_compute_kernel(program)

t = kali.Timer()

# When running in benchmark mode amortize overheads over many runs to measure more accurate GPU times
num_iters = 1000 if args.benchmark else 1

# Setup query pool to measure GPU time
queries = device.create_query_pool(kali.QueryType.timestamp, num_iters * 2)

# Compress!
command_buffer = device.create_command_buffer()
for i in range(num_iters):
    command_buffer.write_timestamp(queries, i * 2)
    encoder.dispatch(
        thread_count=[w, h, 1],
        vars={
            "input_tex": input_tex,
            "decoded_tex": decoded_tex,
            "lr": 0.1,
            "adam_beta_1": 0.9,
            "adam_beta_2": 0.999,
        },
        command_buffer=command_buffer,
    )
    command_buffer.write_timestamp(queries, i * 2 + 1)
command_buffer.submit()

# Wait for GPU to finish and get timestamps
device.wait()

total_cpu_time_sec = t.elapsed_s()

times = np.asarray(queries.get_timestamp_results(0, num_iters * 2))
comp_time_sec = np.mean(times[1::2] - times[0::2])

# Calculate and print performance metrics
if args.benchmark:
    textures_per_sec = 1 / comp_time_sec
    giga_texels_per_sec = w * h * textures_per_sec / 1e9
    print(f"Benchmark:")
    print(f"- Number of optimization steps: {args.opt_steps}")
    print(f"- Compression time: {1e3 * comp_time_sec:.4g} ms")
    print(f"- Compression throughput: {giga_texels_per_sec:.4g} GTexels/s")
    print(f"- Total CPU time: {total_cpu_time_sec:.4g} s")

# Calculate and print PSNR
decoded = decoded_tex.to_numpy()
mse = np.mean((input - decoded) ** 2)
psnr = 20 * np.log10(1.0 / np.sqrt(mse))
print(f"PSNR: {psnr:.4g}")

# Show decoded texture in tev
if args.tev:
    kali.utils.show_in_tev_async(decoded_tex, name="tinybc-decoded")

# Output decoded texture
if args.output_path:
    decoded_tex.to_bitmap().convert(
        pixel_format=kali.Bitmap.PixelFormat.rgb,
        component_type=kali.Bitmap.ComponentType.uint8,
        srgb_gamma=True,
    ).write_async(args.output_path)
